pkg/email/email_test.go

```diff
@@ -0,0 +1,163 @@
+	}
+	subject := "testSubject"
+	receiver := "receiver@test.com"
+	emailer.SendEmail([]string{receiver}, subject, "", nil)
+	if len(d.Message) != 1 {
+		t.Fatalf("message length %d, expected 1", len(d.Message))
+	}
+	tests := []struct {
+		name     string
+		header   string
+		expected string
+	}{
+		{name: "test from header", header: "From", expected: p.Sender},
+		{name: "test To header", header: "To", expected: receiver},
+		{name: "test Subject header", header: "Subject", expected: subject},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			header := d.Message[0].GetHeader(tt.header)[0]
+			if header != tt.expected {
+				t.Errorf("got header: %s, expected header: %s", header, tt.expected)
+			}
+		})
+	}
+}
+
+func TestSendResetPasswordAndInviteUserEmail_RateLimiter(t *testing.T) {
+	// Burst value is set to 0, so any call to SendEmailWithLimit should result in ErrRateLimitExceeded
+	p := email.Params{Burst: 0, LimitEveryDuration: time.Minute}
+	emailer, err := email.NewEmailer(p)
+	testutil.Must(t, err)
+	tests := []struct {
+		name     string
+		op       func(receivers []string, params map[string]string) error
+		receiver []string
+		expect   error
+	}{
+		{name: "invite user limiter", op: emailer.SendInviteUserEmail, receiver: []string{"foo"}, expect: email.ErrRateLimitExceeded},
+		{name: "reset password  limiter", op: emailer.SendResetPasswordEmail, receiver: []string{"foo"}, expect: email.ErrRateLimitExceeded},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := tt.op(tt.receiver, map[string]string{})
+			if !errors.Is(err, email.ErrRateLimitExceeded) {
+				t.Errorf("expected error: %s got error: %s", email.ErrRateLimitExceeded, err)
+			}
+		})
+	}
+}
+
+func TestTemplates(t *testing.T) {
+	p := email.Params{
+		SMTPHost:           "abc@test.com",
+		Sender:             "test@test.com",
+		LimitEveryDuration: time.Minute,
+		Burst:              10,
+	}
+	d := email.Dialer{}
+	lim := rate.NewLimiter(rate.Every(p.LimitEveryDuration), p.Burst)
+	emailer := email.Emailer{
+		Params:  p,
+		Dialer:  &d,
+		Limiter: lim,
+	}
+	m := map[string]string{
+		"paramx": "foo",
+		"paramy": "bar",
+	}
+	receivers := []string{"receiver@test.com"}
+
+	tests := []struct {
+		name         string
+		op           func(receivers []string, params map[string]string) error
+		expectedBody string
+	}{
+		{name: "reset password template", op: emailer.SendResetPasswordEmail, expectedBody: resetPasswordGolden},
+		{name: "invite user template", op: emailer.SendInviteUserEmail, expectedBody: inviteUserGolden},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := tt.op(receivers, m)
+			testutil.Must(t, err)
+			var builder strings.Builder
+			if len(d.Message) != 1 {
+				t.Fatalf("message length %d, expected 1", len(d.Message))
+			}
+			_, err = d.Message[0].WriteTo(&builder)
+			testutil.Must(t, err)
+			msg := strings.NewReplacer("\r\n", "\n").Replace(builder.String())
```

Comment:
You are doing one replace pattern - use `strings.ReplaceAll`, you don't need a replacer which support multiple patterns replace